sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -521,38 +524,138 @@ class Analyzer(
+              val newAggregateExpressions = a.aggregateExpressions ++
+                findNotResolvedMissingAttrs(
+                  a.aggregateExpressions, a.inputSet, missingResolvableAttrs)
+              a.copy(aggregateExpressions = newAggregateExpressions)
+            case o => o
+          }
+
+          // Add missing attributes and then project them away after the sort.
+          Project(child.output,
+            Sort(newOrdering, s.global, newChild))
+        }
+    }
+
+    private def findNotResolvedMissingAttrs(
+        outputSet: AttributeSet,
+        inputSet: AttributeSet,
+        missingAttrs: Seq[Attribute]): Seq[Attribute] = {
+      val resolvedAttrs =
+        missingAttrs.filter(attr => inputSet.exists(_.semanticEquals(attr)))
+      resolvedAttrs.filterNot(attr => outputSet.exists(_.semanticEquals(attr)))
+    }
+
+    private def findNotResolvedMissingAttrs(
+        outputSet: Seq[Expression],
+        inputSet: AttributeSet,
+        missingAttrs: Seq[Attribute]): Seq[Attribute] = {
+      val resolvedAttrs =
+        missingAttrs.filter(attr => inputSet.exists(_.semanticEquals(attr)))
+      resolvedAttrs.filterNot(attr => outputSet.exists(_.semanticEquals(attr)))
+    }
+
+    /**
+     * Traverse the tree until resolving the sorting attributes
+     * Return all the resolvable missing sorting attributes
+     */
+    @tailrec
+    private def collectResolvableMissingAttrs(
+        ordering: Seq[SortOrder],
+        plans: mutable.Queue[LogicalPlan]): (Seq[SortOrder], Seq[Attribute]) = {
+      if (plans.isEmpty) (Seq.empty[SortOrder], Seq.empty[Attribute])
+      else {
+        plans.dequeue() match {
+          // Only Windows and Project have projectList-like attribute.
+          case un: UnaryNode if un.isInstanceOf[Project] || un.isInstanceOf[Window] =>
+            val (newOrdering, missingAttrs) = resolveAndFindMissing(ordering, un, un.child)
+            // If missingAttrs is non empty, that means we got it and return it;
+            // Otherwise, continue to traverse the tree.
+            if (missingAttrs.nonEmpty) (newOrdering, missingAttrs)
+            else {
+              plans.enqueue(un.child)
+              collectResolvableMissingAttrs(ordering, plans)
+            }
+          // Jump over the following UnaryNode types
+          // The output of these types is the same as their child's output
+          case un: UnaryNode
+            if un.isInstanceOf[Distinct] ||
+              un.isInstanceOf[Filter] ||
+              un.isInstanceOf[Limit] ||
+              un.isInstanceOf[RedistributeData] ||
+              un.isInstanceOf[Repartition] ||
+              un.isInstanceOf[RepartitionByExpression] ||
+              un.isInstanceOf[Sample] ||
+              un.isInstanceOf[Sort] ||
+              un.isInstanceOf[SortPartitions] ||
+              un.isInstanceOf[Subquery] ||
+              un.isInstanceOf[With] ||
+              un.isInstanceOf[WithWindowDefinition] =>
```

Comment:
It's correct to list all these here, but I can't image a case that those except Filter are needed.

Subquery/With/WithWindowDefinition/Limit should be part of subquery, How to use use Distinct/RedistributeData/Repartition/RepartitionByExpression in SQL query? For DataFrame, we may not support adding missing attributes for Sort, it's a little bit confusing.